File: src/tools/write.ts
import fs from "node:fs/promises";
import path from "node:path";
import type { ToolResult } from "./bash.js";

export interface WriteParams {
  path: string;
  content: string;
}

export async function write(params: WriteParams): Promise<ToolResult> {
  try {
    await fs.mkdir(path.dirname(params.path), { recursive: true });
    await fs.writeFile(params.path, params.content, "utf-8");
    return { content: `Successfully wrote to ${params.path}` };
  } catch (err: any) {
    return {
      content: `Error writing file: ${err.message}`,
      isError: true,
    };
  }
}

File: pincer/tools/write.go
package tools

import (
	"fmt"
	"os"
	"path/filepath"
)

type WriteParams struct {
	Path    string `json:"path"`
	Content string `json:"content"`
}

func Write(p *WriteParams) (*Result, error) {
	dir := filepath.Dir(p.Path)
	if err := os.MkdirAll(dir, 0755); err != nil {
		return &Result{Content: "Error creating directory: " + err.Error(), IsError: true}, nil
	}

	if err := os.WriteFile(p.Path, []byte(p.Content), 0644); err != nil {
		return &Result{Content: "Error writing file: " + err.Error(), IsError: true}, nil
	}

	return &Result{Content: fmt.Sprintf("Successfully wrote to %s", p.Path)}, nil
}
